segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -552,36 +559,173 @@ private CompletableFuture<Void> writeSnapshotInfo(long snapshotId) {
+                                    }
+                                    return null;
+                                }, executor)
+                                .handleAsync((v, e) -> {
+                                    if (null != e) {
+                                        val ex = Exceptions.unwrap(e);
+                                        if (ex instanceof StorageNotPrimaryException) {
+                                            log.warn("SystemJournal[{}] Error while saving snapshot info to LTS {}. info = {}", containerId, snapshotInfoFileName, info, e);
+                                            throw new CompletionException(e);
+                                        }
+                                        if (attempts.incrementAndGet() > config.getMaxJournalWriteAttempts()) {
+                                            log.warn("SystemJournal[{}] Error while saving snapshot info to LTS. File = {}. info = {}", containerId, snapshotInfoFileName, info, e);
+                                            throw new CompletionException(e);
+                                        }
+                                        log.warn("SystemJournal[{}] Error while saving snapshot info to LTS. File = {}. info = {}", containerId, snapshotInfoFileName, info, e);
+                                    } else {
+                                        // No exception we are done
+                                        log.info("SystemJournal[{}] Snapshot info saved successfully. info = {}", containerId, info);
+                                        isDone.set(true);
+                                    }
+                                    return null;
+                                }, executor);
+                    }, executor);
+
+        } catch (IOException e) {
+            return CompletableFuture.failedFuture(new ChunkStorageException(snapshotInfoFileName, "Unable to serialize", e));
+        }
+    }
+
     /**
      * Find and apply latest snapshot.
      */
     private CompletableFuture<SystemSnapshotRecord> findLatestSnapshot() {
         // Step 1: Read snapshot info.
+        return readSnapshotInfo()
+            .thenComposeAsync(snapshotInfo -> {
+                if (null != snapshotInfo) {
+                    val snapshotFileName = NameUtils.getSystemJournalSnapshotFileName(containerId, snapshotInfo.getEpoch(), snapshotInfo.getSnapshotId());
+                    log.debug("SystemJournal[{}] Snapshot info read. {} pointing to {}", containerId, snapshotInfo, snapshotFileName);
+
+                    // Step 2: Validate and Read contents.
+                    return getContents(snapshotFileName, false)
+                            // Step 3: Deserialize and return.
+                            .thenApplyAsync(contents -> readSnapshotRecord(snapshotInfo, contents), executor)
+                            .exceptionally(e -> {
+                                throw new CompletionException(new IllegalStateException(
```

Comment:
This is a very strange conversion. Its taking any exception that occurs and turning it into an IllegalState. Why not simply propagate the original exception? How do you know it is an illegal state?

Reply:
Storage Metadata about all segments is stored in storage metadata table. And metadata about this special table segments is stored in journals. Periodically we snapshot and truncate these journals.
On restart, at the very start of segment store, we replay these system journals to recreate the state of storage metadata segment starting from the latest snapshot. 
Snapshot info read in these lines points from what journal to start replaying. This information is stored in BK and now also in a special chunk on LTS. 
If snapshot pointed by snapshot info cannot be read (for whatever reason) then it is indeed an illegal state. Container cannot start if this the case.
(As it was already validated that it can be read back at time when we saved it first.